File: packages/core-native/src/host/KitbagHostObject.ts
// The JSI HostObject surface — the polled realtime read path (SPEC §13.2, §13.3).
//
// A C++ jsi::HostObject (packages/core-native/cpp/KitbagHostObject.cpp) is
// installed ONCE on the JS runtime's global. It holds the single kb_engine*
// (SPEC §4.5) and is the only holder; the sibling TurboModule (commands) reaches
// the engine through it, not by owning its own pointer. Each method calls
// straight into the C ABI, synchronously, with zero serialisation — the direct
// analogue of the Dart FFI call it replaces.
//
// §13.3, the rule that matters: these reads are what a Reanimated worklet polls
// on the UI thread each frame — reading `global.__KitbagHostObject.bar_phase`
// (and its siblings) as a plain property, writing a SharedValue. They are
// number-valued PROPERTIES, not methods: a JSI double property is allocation-free
// per access, whereas a callable would allocate a jsi::Function per frame, which
// §13.3 forbids on the 60fps path. They must NEVER be routed through React state
// or a JS-thread poll. This file is only the typed contract; it deliberately
// holds no state and starts no loop.
//
// This is a SKELETON. There is no native build wired until #31, so nothing is
// installed on `global` at runtime yet — getKitbagHostObject() throws until then.
// The C++ that fulfils this contract is uncompiled scaffolding (see cpp/).

/**
 * Synchronous polled reads into the C ABI. Every read is a `number`-valued
 * PROPERTY (JSI double), not a method: reading it re-invokes the underlying C ABI
 * call in C++ `get()` and returns the value directly, allocating nothing per
 * access — the 60fps requirement (§13.3). The engine's u64/i64 frame counts are
 * exact to 2^53 (~5,900 years at 48kHz), so no BigInt crosses (SPEC §13.2).
 */
export interface KitbagHostObject {
  /** kb_metronome_bar_phase — position within the bar, [0, 1). Beat sweep. */
  readonly bar_phase: number;
  /** kb_metronome_current_beat — beat index within the bar, -1 when stopped. */
  readonly current_beat: number;
  /** kb_metronome_current_bpm — effective BPM including ramp progress. */
  readonly current_bpm: number;
  /** kb_engine_frames_rendered — monotonic master clock, in frames. */
  readonly frames_rendered: number;
  /** kb_tuner_snapshot — packed reading; decode with src/host/snapshot.ts. */
  readonly tuner_snapshot: number;
  /** kb_player_position — single-source transport position, in frames. */
  readonly player_position: number;
}

/**
 * The value `current_beat` carries while the metronome is stopped, per the
 * kb_metronome_current_beat contract. One owner: every LED row that seeds or
 * compares a beat index reads it from here rather than retyping -1 (§13.7).
 */
export const KB_STOPPED_BEAT = -1;

/**
 * The global key the C++ installer publishes the HostObject under. One owner of
 * this string: TS setup validates via it, and the worklet read path reads the
 * same key on the UI thread.
 */
export const KITBAG_HOST_OBJECT_KEY = '__KitbagHostObject';

interface HostGlobal {
  [KITBAG_HOST_OBJECT_KEY]?: KitbagHostObject;
}

/**
 * Fetch the installed HostObject, or throw if the native module has not
 * installed it. For JS-thread setup/validation only — the 60fps path reads the
 * `global.__KitbagHostObject.bar_phase` property directly inside a worklet
 * (SPEC §13.3), never through this JS-thread accessor.
 */
export function getKitbagHostObject(): KitbagHostObject {
  const host = (globalThis as HostGlobal)[KITBAG_HOST_OBJECT_KEY];
  if (host === undefined) {
    throw new Error(
      `${KITBAG_HOST_OBJECT_KEY} is not installed. The core-native module must ` +
        'install the JSI HostObject before any read (SPEC §13.2). No native ' +
        'build is wired yet (#31).',
    );
  }
  return host;
}

File: packages/core-native/cpp/KitbagHostObject.cpp
// Implementation of the JSI HostObject (SPEC §13.2).
//
// STATUS: SKELETON (#31) — see KitbagHostObject.h. The native build is wired and
// this compiles/links against kitbag_core, but it has not run inside a live RN
// runtime; that install and its on-device measurement are #33.
//
// The six reads are exactly the polled realtime set (SPEC §13.2, §13.3):
// bar_phase, current_beat, current_bpm, frames_rendered, tuner_snapshot,
// player_position. Each is a number-valued PROPERTY, not a method: get()
// re-invokes the matching kb_* call and returns its value directly, so a worklet
// reading global.__KitbagHostObject.bar_phase each frame on the UI thread
// allocates nothing — no jsi::Function per access, which §13.3 forbids on the
// 60fps path.

#include "KitbagHostObject.h"

#include <string>

#include "kitbag_api.h"

namespace kitbag {

using facebook::jsi::PropNameID;
using facebook::jsi::Runtime;
using facebook::jsi::Value;

Value KitbagHostObject::get(Runtime& rt, const PropNameID& name) {
  const std::string prop = name.utf8(rt);
  kb_engine* engine = engine_;

  if (prop == "bar_phase") {
    return Value(static_cast<double>(kb_metronome_bar_phase(engine)));
  }
  if (prop == "current_beat") {
    return Value(static_cast<double>(kb_metronome_current_beat(engine)));
  }
  if (prop == "current_bpm") {
    return Value(static_cast<double>(kb_metronome_current_bpm(engine)));
  }
  if (prop == "frames_rendered") {
    return Value(static_cast<double>(kb_engine_frames_rendered(engine)));
  }
  if (prop == "tuner_snapshot") {
    // uint64, 48 bits used — exact as a double, decoded in src/host/snapshot.ts.
    return Value(static_cast<double>(kb_tuner_snapshot(engine)));
  }
  if (prop == "player_position") {
    return Value(static_cast<double>(kb_player_position(engine)));
  }
  return Value::undefined();
}

std::vector<PropNameID> KitbagHostObject::getPropertyNames(Runtime& rt) {
  std::vector<PropNameID> names;
  for (const char* prop :
       {"bar_phase", "current_beat", "current_bpm", "frames_rendered",
        "tuner_snapshot", "player_position"}) {
    names.push_back(PropNameID::forUtf8(rt, prop));
  }
  return names;
}

void installKitbagHostObject(Runtime& rt, kb_engine* engine) {
  auto host = std::make_shared<KitbagHostObject>(engine);
  rt.global().setProperty(
      rt, kHostObjectKey, Value(rt, facebook::jsi::Object::createFromHostObject(rt, host)));
}

}  // namespace kitbag
